common/thorhelper/roxiehelper.cpp

```diff
@@ -1533,51 +1533,64 @@ bool CSafeSocket::readBlocktms(StringBuffer &ret, unsigned timeoutms, HttpHelper
                 pHttpHelper->setHttpMethod(HttpMethod::GET);
         }
 
+        bool sockClosed = false;
         if (pHttpHelper && pHttpHelper->isHttp())
         {
 #define MAX_HTTP_HEADERSIZE 16000 //arbitrary per line limit, most web servers are lower, but REST queries can be complex..
             char header[MAX_HTTP_HEADERSIZE + 1]; // allow room for \0
-            sock->readtms(header, 1, MAX_HTTP_HEADERSIZE, bytesRead, timeoutms);
-            header[bytesRead] = 0;
-            char *payload = strstr(header, "\r\n\r\n");
-            if (payload)
+            const char *val = nullptr;
+            unsigned readtimeout;
+            if (timeoutms == INFINITE)
+                readtimeout = 60000;
+            else
+                readtimeout = timeoutms;
+            CTimeMon tm(readtimeout);
+            unsigned totalRead = 0;
+            unsigned remaining;
+            len = 0;
+            while (!sockClosed && !tm.timedout(&remaining))
             {
-                *payload = 0;
-                payload += 4;
+                unsigned maxPoss = MAX_HTTP_HEADERSIZE - totalRead;
+                sockClosed = readtmsAllowClose(sock, &header[totalRead], 1, maxPoss, bytesRead, remaining);
```

Comment:
(I thought I'd commented on this before, but lost..)

What happens if reads N bytes that include the "\r\n\r\n", but not the whole payload?.
I think the code as it stands would read whatever it had got so far (irrespective of timeout)
Shouldn't it wait and continue to read the complete Content-Length, or timeout?